### PR TITLE
Makefile.include: change to account for TOOLCHAIN having dependencies

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

While looking at #15828 I realized a couple things:

- `TOOLCHAIN` is used to trigger dependencies (for `picolibc` in `gnrc_minimal`)
- `TOOLCHAIN` is not saved in the dependency debug variables
- It is not parsed after `BOARD` dependencies are handler, e.g.: if the `BOARD` is `native` on OSX the default toolchain would be `llvm` but `picolibc` is not added.

IMO if  `TOOLCHAIN` can have dependencies as proposed by @aabadie  in #15828 and as used in `examples/gnrc_minimal` then the above changes are required, but TBH I would prefer `TOOLCHAIN` not having dependencies.

### Testing procedure

I don't think moving `TOOLCHAIN` up can have an impact, but please take a close look.

`info-board-supported` now has consistent results same as `info-debug-variable-FEATURES_OPTIONAL` on `gnrc_minimal`.

### Issues/PRs references

Depends on #15828

### CODE REVIEW

**Comment by fjmolinas** [comment]
ups wrong label

File: examples/gnrc_minimal/Makefile.board.dep [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# llvm compatibility requires yet to be released picolibc 1.4.7
```
Comment (aabadie) [inline examples/gnrc_minimal/Makefile.board.dep:1]: You could use a `Makefile.gcc.dep` and in the main Makefile.dep, add `include -$(APPDIR)/Makefile.$(TOOLCHAIN).dep`. It would be cleaner than reuse a file that is meant for board specific dependencies. What do you think ?

File: pkg/libfixmath/Makefile.dep [hunk 2]
```diff
@@ -6,3 +6,9 @@ ifneq (,$(filter libfixmath-unittests,$(USEMODULE)))
   # The round function is not provided by the msp430 toolchain
   FEATURES_BLACKLIST += arch_msp430
 endif
+
+ifneq (,$(filter gnu gcc,$(TOOLCHAIN)))
+  # On recent arm-none-eabi-gcc toolchains division testcases fail
+  # see #14568
```
Comment (aabadie) [inline pkg/libfixmath/Makefile.dep:12]: Maybe put the url instead of the issue number.

File: examples/gnrc_minimal/Makefile.board.dep (hunk 1 above)
Comment (fjmolinas) [reply examples/gnrc_minimal/Makefile.board.dep:1]: Sure!

**Comment by fjmolinas** [comment]
Removed commit in#15828

**Comment by fjmolinas** [comment]
@aabadie Murdock is green. Should I squash? I think moving `TOOLCHAIN` earlier should not have any impact, 4 makefiles are included before:

- `info.inc.mk`: two recipes depend on the TOOLCHAIN `info-toolchains-supported`, `check-toolchain-supported`. The handling of the first is not affected, while the second is now properly checked
- `scan-build.inc.mk`: only works with llvm, the `TOOLCHAIN` is checked and set in the recipes
- `info-global.inc.mk`: will handle `TOOLCHAIN` with this change
- `buildtests.inc.mk`: uses `check-toolchain-supported`.

File: examples/gnrc_minimal/Makefile.gcc.dep [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# llvm compatibility requires yet to be released picolibc 1.4.7
+ifneq (,$(filter gnu,$(TOOLCHAIN)))
```
Comment (aabadie) [inline examples/gnrc_minimal/Makefile.gcc.dep:2]: You don't need this ifneq now. This file is included only if the gcc toolchain is used :)

File: Makefile.include [hunk 4]
```diff
@@ -173,6 +173,27 @@ include $(RIOTMAKE)/boards.inc.mk
...
+endif
+# Use override so that we can redefine a variable set on the command line (as
+# opposed to one set in the environment)
+ifeq (clang,$(TOOLCHAIN))
+  # TOOLCHAIN = clang is an alias for TOOLCHAIN = llvm
+  override TOOLCHAIN := llvm
+endif
+ifeq (gcc,$(TOOLCHAIN))
+  # TOOLCHAIN = gcc is an alias for TOOLCHAIN = gnu
+  override TOOLCHAIN := gnu
```
Comment (aabadie) [inline Makefile.include:191]: Does this means that TOOLCHAIN is equal to `gnu` when Makefile.dep is processed ? If yes, the toolchain specific dependency file should be called `Makefile.gnu.dep` and not `Makefile.gcc.dep` (in  examples/gnrc_minimal) ?

Comment (fjmolinas) [reply Makefile.include:191]: True!

**Review by aabadie — DISMISSED** [review]
LGTM.

ACK, please squash!

**Comment by fjmolinas** [comment]
Well I bodged something

**Comment by aabadie** [comment]
@fjmolinas I was too sure of myself, so I (force) pushed a squashed branch to this PR. Unfortunately that doesn't work either for non ARM targets. Maybe we should exclude picolibc for non ARM targets and use FEATURES_REQUIRED instead of FEATURES_OPTIONAL ?

**Comment by fjmolinas** [comment]
I'm trying something new...model newlib as a `FEATURE` as well...

**Comment by fjmolinas** [comment]
> I'm trying something new...model newlib as a `FEATURE` as well...

If ti works I'll split out that commit.

File: makefiles/features_modules.inc.mk [hunk 5]
```diff
@@ -45,5 +45,15 @@ ifneq (,$(filter cortexm_mpu,$(FEATURES_USED)))
...
 
+# use picolibc if the feature is used
+ifneq (,$(filter picolibc,$(FEATURES_USED)))
+  USEMODULE += picolibc
+endif
+
+# use newlib if the feature is used
+ifneq (,$(filter newlib,$(FEATURES_USED)))
+  USEMODULE += newlib
+endif
```
Comment (fjmolinas) [inline makefiles/features_modules.inc.mk:56]: I should probably add a `FEATURES_CONFLICTING` as well.

Comment (fjmolinas) [reply makefiles/features_modules.inc.mk:56]: For cases where both are provided

File: cpu/cortexm_common/Makefile.dep [hunk 6]
```diff
@@ -4,21 +4,13 @@ USEMODULE += cortexm_common
...
-else
-  # all cortex MCU's use newlib as libc
-  USEMODULE += newlib
+FEATURES_REQUIRED_ANY += newlib|picolibc
+ifneq (,$(filter newlib,$(USEMODULE)))
   # use the nano-specs of Newlib when available
   USEMODULE += newlib_nano
 endif
 
+
```
Comment (aabadie) [inline cpu/cortexm_common/Makefile.dep:13]: unrelated

**Comment by aabadie** [comment]
Hmm, you forgot Kconfig :)

**Comment by fjmolinas** [comment]
> Hmm, you forgot Kconfig :)

At least everything else seems fine, should I split into a different PR?

**Comment by aabadie** [comment]
> should I split into a different PR?

I don't think that needed. The change seems reasonable. Just needs some more testing.

**Comment by fjmolinas** [comment]
> > should I split into a different PR?
> 
> I don't think that needed. The change seems reasonable. Just needs some more testing. You could change the PR title.

I think I'll split it will make testing easier.

**Comment by fjmolinas** [comment]
@aabadie May I squash? The old commits had already been reviewed, I removed the ones that where split off.

**Comment by aabadie** [comment]
Please squash!

**Review by aabadie — APPROVED** [review]
Now that newlib is modeled as a feature, there's no more build issue with this PR.

I agree with the idea of this PR and code changes are still fine.

ACK
